### DISCUSSION TITLE
Support for Linux?

### DISCUSSION THREAD

**Discussion by beaussan** [johnlindquist/kit#121] [Ideas]
Hi, this project looks amazing!

I was wondering if supports Linux or plan to. I didn't find any good Linux equivalent to what you are building and I would love to remplace my own hacky rofi scripts with this!

This project looks soo promising!

**Comment by johnlindquist** [comment]
Yes, it's in the plans. I don't foresee it happening until later this year though:

1. I need to stabilize a final mac release
2. I would play a support role for someone who wants to do the port of the app. The main issues I foresee would be around the app prompt focus/position/etc, the auto-updater, and maybe shortcuts. It's possible it would _mostly_ work if I just did a build today.
3.  The community would have to handle the kits for all the os-specifc libs. Window management, audio, etc. They're in AppleScript now, but I'm sure Linux would have equivalent tools we could hook into.
4. I can handle the build/distribution/etc

**Reply by beaussan** [reply]
Hey! Congrats on the release!

I wanted to know if the Linux version is still planed, I know the cumminity size is smaller than mac, but I would love to use it!

Thanks =)

**Reply by johnlindquist** [reply]
Yes, I'll need to set up a Linux machine next year. I know there's a couple of install/startup steps that need to be customized per platform, but everything else "should" be good to go.

**Comment by johnlindquist** [comment]
As a general update here, I've started publishing Linux builds:

https://github.com/johnlindquist/kitapp/releases/latest

These are untested (I don't have a Linux machine yet), so I'd need community support if we run into any errors.

Note: I have a big refactor in the `vite` branch that I'll be moving over the `kit` repo as a monorepo that will support more Linux distros and make it waaaay, waaaaaaay easier to contribute to dev when we run into issues.

**Reply by canvural** [reply]
I'll try to test it this week. Thank you!

**Comment by takanome-dev** [comment]
Hey @johnlindquist thanks for building such an amazing app. Keep up the good work.

I'm having a hard time to install the app. For example, I didn't know where to run `npm run build-kit` by following the docs because the docs don't mention where that script should be run (cd kit/kitapp and run the script). I figured it out after looking into the `package.json`.

Then, I `cd` into `kit`, ran the script and get this error:
```sh
Error: Command failed with exit code 2: npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"
node_modules/@octokit/core/node_modules/@octokit/openapi-types/types.d.ts(20302,7): error TS1005: '}' expected.
    at makeError (file:///home/takanome_dev/Downloads/Programs/kit/node_modules/@johnlindquist/globals/dist/index.js:28:111641)
    at file:///home/takanome_dev/Downloads/Programs/kit/node_modules/@johnlindquist/globals/dist/index.js:28:123014
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 3)
    at async file:///home/takanome_dev/Downloads/Programs/kit/build/build-kit.js:77:1 {
  shortMessage: 'Command failed with exit code 2: npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"',
  command: 'npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"',
  escapedCommand: 'npx tsc --project "./tsconfig-cjs.json" --outDir "\\"/home/takanome_dev/.kit/cjs\\""',
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: "node_modules/@octokit/core/node_modules/@octokit/openapi-types/types.d.ts(20302,7): error TS1005: '}' expected.",
  stderr: '',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
```

I also downloaded  `Kit-SDK-Linux-1.40.62-x64.tar.gz ` and ran `npm run build-kit` and get this error:
```sh
Error: Command failed with exit code 1: npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"
error TS5058: The specified path does not exist: './tsconfig-cjs.json'.
    at makeError (file:///home/takanome_dev/Downloads/Programs/Kit-SDK-Linux-1.40.62-x64/node_modules/@johnlindquist/globals/dist/index.js:28:111641)
    at file:///home/takanome_dev/Downloads/Programs/Kit-SDK-Linux-1.40.62-x64/node_modules/@johnlindquist/globals/dist/index.js:28:123014
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 3)
    at async file:///home/takanome_dev/Downloads/Programs/Kit-SDK-Linux-1.40.62-x64/build/build-kit.js:77:1 {
  shortMessage: 'Command failed with exit code 1: npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"',
  command: 'npx tsc --project ./tsconfig-cjs.json --outDir "/home/takanome_dev/.kit/cjs"',
  escapedCommand: 'npx tsc --project "./tsconfig-cjs.json" --outDir "\\"/home/takanome_dev/.kit/cjs\\""',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: "error TS5058: The specified path does not exist: './tsconfig-cjs.json'.",
  stderr: '',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false

```

**Comment by johnlindquist** [comment]
@TAKANOME-DEV

# The App Setup Process in Detail

You can see the code for the following steps here. The app attempts to do them automatically, but it obviously currently failing for Linux users: https://github.com/johnlindquist/kitapp/blob/main/src/main.dev.ts#L845

## The 3 Required Directories

### 1. `~/.kit`

This is the Kit SDK. You _should_ be able to download the SDK and extract it here. (This is also what `npm run build` tries to build)

https://github.com/johnlindquist/kitapp/releases/latest

> Note - The SDK is just a built version of this repo with all the deps bundled to save on initial setup.

You should extract it so the package.json resides here: `~/.kit/package.json`

### 2. `~/.kenv`

This is the Kit Environent where scripts are stored. It's simply a way to organize scripts, logs, tmp, etc. You can clone it from: 
https://github.com/johnlindquist/kenv

You should clone it so the package.json resides here: `~/.kenv/package.json`

### 3. `~/.knode`

This is a version of node (v16.17.1) that matches the current version of the app. 

https://nodejs.org/download/release/v16.17.1/

`node` should be accessible as `~/.knode/bin/node`

## Final Configuration Setups

### Intall `esbuild` for TypeScript support

Use the .knode version of node to install esbuild into the ~/.kit dir:
```
cd ~/.kit && PATH=~/.knode/bin:$PATH npm i esbuild@0.16.15 --save-exact
```

### Run the setup scripts

```
cd ~/.kit && ~/.knode/bin/node ~/.kit/run/terminal.js ~/.kit/setup/setup.js
```

## Thoughts on the Setup Process
- `esbuild` is a different binary per-platform. It's been a challenge to get the CI process bundling the correct version and fighting against code-signing of unsign binaries. I an _not_ an expert in this area.
- Keeping the `~/.kit` and `~/.knode` SDK outside of the app makes it possible to run scripts from the terminal and allows users to play with the SDK code before submitting PRs. It would be much easier to package it in the app, but we'd lose features. 
- In v2, I'm considering the implications of moving ~/.kit and ~/.knode to the app directory. But haven't come to any conclusions
- There's a huge refactor in process that will clean up all the mess of setting up the dev environment. I hesitate to support the current way too much since it's a mess and it's going away soon.

## Did That Work?

Please let me know if any of those steps fail. I'm pretty flooded with support requests right now after the hackernews frontpage feature, but I'd still love to help Linux users get this working.

**Reply by takanome-dev** [reply]
Thanks for the quick response. Really appreciate it.
The step that fails is running the setup script `cd ~/.kit && ~/.knode/bin/node ~/.kit/setup/setup.js` with the following error:
```sh
file:///home/takanome_dev/.kit/setup/setup.js:1
await setup("create-env");
^

ReferenceError: setup is not defined
    at file:///home/takanome_dev/.kit/setup/setup.js:1:1
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12)
```

Btw, I would love to help and contribute to the project. Don't know anything about Electron yet but I'm willing to dive into it. Any guidance is appreciated :pray:

**Reply by johnlindquist** [reply]
@TAKANOME-DEV 

Sorry, mistyped on my end. Here's the fix:

#### Before
```
cd ~/.kit && ~/.knode/bin/node ~/.kit/setup/setup.js
```

#### After
```
cd ~/.kit && ~/.knode/bin/node ~/.kit/run/terminal.js ~/.kit/setup/setup.js
```

### Explanation

The `~/.kit/run/terminal.js` is what loads the Kit SDK to run scripts in the terminal. `setup` was undefined because the SDK wasn't loaded.

**Reply by takanome-dev** [reply]
Great, it works :+1: 
After running that script, I `cd` into `kitapp` and ran `yarn webpack-dev`:
<details>
```sh
20 warnings have detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

ERROR in ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/utils.js 493:6-26
Module not found: Error: Can't resolve 'uglify-js' in '/home/takanome_dev/Downloads/Programs/kitapp/node_modules/webpack/node_modules/terser-webpack-plugin/dist'
 @ ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/index.js 28:4-22
 @ ./node_modules/webpack/lib/config/defaults.js 1062:25-57
 @ ./node_modules/webpack/lib/index.js 328:10-66
 @ ./node_modules/monaco-editor-webpack-plugin/out/index.js 206:77-95
 @ dll renderer renderer[52]

ERROR in ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/utils.js 571:14-34
Module not found: Error: Can't resolve '@swc/core' in '/home/takanome_dev/Downloads/Programs/kitapp/node_modules/webpack/node_modules/terser-webpack-plugin/dist'
 @ ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/index.js 28:4-22
 @ ./node_modules/webpack/lib/config/defaults.js 1062:25-57
 @ ./node_modules/webpack/lib/index.js 328:10-66
 @ ./node_modules/monaco-editor-webpack-plugin/out/index.js 206:77-95
 @ dll renderer renderer[52]

ERROR in ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/utils.js 661:18-36
Module not found: Error: Can't resolve 'esbuild' in '/home/takanome_dev/Downloads/Programs/kitapp/node_modules/webpack/node_modules/terser-webpack-plugin/dist'
 @ ./node_modules/webpack/node_modules/terser-webpack-plugin/dist/index.js 28:4-22
 @ ./node_modules/webpack/lib/config/defaults.js 1062:25-57
 @ ./node_modules/webpack/lib/index.js 328:10-66
 @ ./node_modules/monaco-editor-webpack-plugin/out/index.js 206:77-95
 @ dll renderer renderer[52]

ERROR in dll renderer renderer[2]
Module not found: Error: Can't resolve '@heroicons/react' in '/home/takanome_dev/Downloads/Programs/kitapp'

ERROR in dll renderer renderer[6]
Module not found: Error: Can't resolve '@types/detect-port' in '/home/takanome_dev/Downloads/Programs/kitapp'

ERROR in dll renderer renderer[7]
Module not found: Error: Can't resolve '@types/tail' in '/home/takanome_dev/Downloads/Programs/kitapp'

ERROR in dll renderer renderer[8]
Module not found: Error: Can't resolve '@vscode/codicons' in '/home/takanome_dev/Downloads/Programs/kitapp'

7 errors have detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.60.0 compiled with 7 errors and 20 warnings in 25613 ms
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>

I also ran `yarn install-electron-deps`:
<details>
```sh
yarn run v1.22.19
$ node -r @babel/register .erb/scripts/CheckNativeDep.js && npx electron-builder install-app-deps
  • electron-builder  version=23.3.3
  • loaded configuration  file=package.json ("build" field)
  • installing production dependencies  platform=linux arch=x64 appDir=/home/takanome_dev/Downloads/Programs/kitapp/src
  ⨯ /home/takanome_dev/.nvm/versions/node/v16.17.1/bin/node process failed ERR_ELECTRON_BUILDER_CANNOT_EXECUTE
Exit code:
1
Output:
[1/4] Resolving packages...
[2/4] Fetching packages...
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

Error output:
error node-mac-permissions@2.2.1: The platform "linux" is incompatible with this module.
error Found incompatible module.
  failedTask=installAppDeps stackTrace=Error: /home/takanome_dev/.nvm/versions/node/v16.17.1/bin/node process failed ERR_ELECTRON_BUILDER_CANNOT_EXECUTE
Exit code:
1
Output:
[1/4] Resolving packages...
[2/4] Fetching packages...
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
                                                                                                                                                                                                                                                                                                                                                                                                                                 Error output:
error node-mac-permissions@2.2.1: The platform "linux" is incompatible with this module.
error Found incompatible module.
                                                                                                                                                                                                                                                                                                                                                                                                                                     at ChildProcess.<anonymous> (/home/takanome_dev/Downloads/Programs/kitapp/node_modules/builder-util/src/util.ts:250:14)
    at Object.onceWrapper (node:events:628:26)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1093:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>
Some missing packages (even after installing the packages) and platform incompatibility issues :cry:

**Reply by johnlindquist** [reply]
@TAKANOME-DEV Oh, sorry, I meant you should be able to open the Appimage now. That was the setup process for the production release

**Reply by takanome-dev** [reply]
Shoot, thanks :+1: 
FYI, I downloaded [Kit-SDK-Linux-1.40.62-x64.tar.gz](https://github.com/johnlindquist/kitapp/releases/download/v1.40.62/Kit-SDK-Linux-1.40.62-x64.tar.gz). I have no clue how to start the app, I don't see it in my apps and `yarn start` give me an error.
Sorry for bothering you with this :disappointed_relieved:

**Reply by johnlindquist** [reply]
@TAKANOME-DEV use the AppImage based on your arch:

https://github.com/johnlindquist/kitapp/releases/tag/v1.40.62

![CleanShot 2023-01-23 at 08 58 03@2x](https://user-images.githubusercontent.com/36073/214086487-1c84a53e-aaf5-46d0-9793-1de0bb0e51e4.png)

**Reply by takanome-dev** [reply]
Thanks, I'm using ubuntu but I saw that It's possible to run AppImage on ubuntu. I'm going to give it a try.
Thanks for the support :pray:

**Reply by mstine** [reply]
Can confirm this is all working for me now.

I am seeing some weird font artifacts in the GUI. I'd previously seen these before in Logseq after an upgrade, and the temporary solution was to pass the `--disable-gpu` argument inside the `.desktop` file, which I assume was ultimately being consumed by Electron (way outside my area of expertise). Where would I do this in Kit?

**Reply by johnlindquist** [reply]
@mstine I think you would just launch the AppImage from your terminal and pass in the `--disable-gpu` flag when launching. Sorry, this isn't my area of expertise.

Something like this maybe?

```bash
./Kit.AppImage --disable-gpu
```

I'll look into adding a "disableGpu" option from the menubar

**Reply by mstine** [reply]
That indeed does the trick. I didn't realize the AppImage was still in play post-setup. 

By the way, the menubar doesn't do anything but sit there and look pretty in my Pop!_OS 22.04 LTS environment. But I didn't even look for it until you mentioned it in your last reply. 

I think I'll add a startup item to execute `$HOME/bin/Kit.AppImage --disable-gpu` just to keep things simple.

**Reply by johnlindquist** [reply]
@mstine So clicking/right-clicking on the menubar does nothing?

**Reply by mstine** [reply]
Correct.

**Reply by johnlindquist** [reply]
@mstine Hrmmmm, time to check the Electron issues for Linux tray support 🤔

**Comment by takanome-dev** [comment]
The [Kit-Linux-1.40.62-arm64.AppImage](https://github.com/johnlindquist/kitapp/releases/download/v1.40.62/Kit-Linux-1.40.62-arm64.AppImage) don't work for some reason with an `exec format error` but [Kit-Linux-1.40.62-x86_64.AppImage](https://github.com/johnlindquist/kitapp/releases/download/v1.40.62/Kit-Linux-1.40.62-x86_64.AppImage) does work :rocket: 

The issue I'm facing, though, is that it works the first time you install and run it but, if you start it later, you will see the icon on the gnome bar but the app won't show up :thinking:

**Reply by johnlindquist** [reply]
Hmm, I've been working on the CI (GitHub workflows) that build the app, hopefully I'll be able to solve the arm64 version that way.

> if you start it later

Can you make sure that when you quit, there's no `Kit` processes hanging around. The app starts a couple of child processes which should end when quit.

Also, `cmd+0` (maybe ctrl+0 on Linux?) should empty the prompt cache (`~/.kit/db/prompt.json`) which is what stores the position of the prompt if you move it around. If your displays change (new monitor, etc) sometimes the prompt position will get confused. Probably not the issue, but worth checking.

**Reply by takanome-dev** [reply]
I killed the process, manually cleanup `~/.kit/db/prompt.json` to `{}` and start kit using the terminal:

![kit-1](https://user-images.githubusercontent.com/79809121/214135358-b61ea342-6f39-46d1-b0e9-bcf6501d1ab3.png)

**Reply by johnlindquist** [reply]
@TAKANOME-DEV Did that work to help it re-open after quit? I don't understand what you're showing me in the screenshot

**Reply by takanome-dev** [reply]
You can see that it's running but the UI is not here (is not showed up) which is weird.

**Reply by johnlindquist** [reply]
@TAKANOME-DEV If you right-click the menubar icon -> Debug -> Open Dev Tools

Does anything show up?

Does pressing the main keyboard shortcut ctrl+; make anything happen in the logs?

**Reply by takanome-dev** [reply]
Nothing happens when I press the shortcut or I click the kit icon in the menubar. But I found errors in the logs:

![kit-2](https://user-images.githubusercontent.com/79809121/214164029-aed26fd2-74f6-4fc2-bd50-e4b5596c543b.png)
![kit-3](https://user-images.githubusercontent.com/79809121/214164062-57c9828f-280b-4551-a325-1010f29f8618.png)
```sh
22:18:25.906 › 🎨 Creating tray... { checkDb: true }
22:18:25.908 › 🎨 Tray icon: busy
22:18:25.954 › Tray created
22:18:25.963 › 🧐 Checking /home/takanome_dev/.kit {
  cwd: '/home/takanome_dev/.kit',
  encoding: 'utf-8',
  env: {
    KIT: '/home/takanome_dev/.kit',
    KENV: '/home/takanome_dev/.kenv',
    PATH: '/home/takanome_dev/.knode/bin:/home/takanome_dev/.kit/bin:/home/takanome_dev/.kit/override/code:/home/takanome_dev/.kenv/bin:/tmp/.mount_Kit-Libizfr6:/tmp/.mount_Kit-Libizfr6/usr/sbin:/home/takanome_dev/.local/share/pnpm:/home/takanome_dev/.nvm/versions/node/v18.8.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/tmp/.mount_Kit-Libizfr6:/tmp/.mount_Kit-Libizfr6/usr/sbin:/home/takanome_dev/.local/share/pnpm:/home/takanome_dev/.nvm/versions/node/v18.8.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/takanome_dev/.kit/bin'
  },
  stdio: 'pipe'
}
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
Error: Bad atom
    at ReadFixedRequest.callback (/tmp/.mount_Kit-Libizfr6/resources/app.asar/node_modules/x11/lib/xcore.js:465:29)
    at ReadFixedRequest.execute (/tmp/.mount_Kit-Libizfr6/resources/app.asar/node_modules/x11/lib/unpackstream.js:41:10)
    at UnpackStream.resume (/tmp/.mount_Kit-Libizfr6/resources/app.asar/node_modules/x11/lib/unpackstream.js:165:30)
    at UnpackStream.write (/tmp/.mount_Kit-Libizfr6/resources/app.asar/node_modules/x11/lib/unpackstream.js:102:10)
    at Socket.<anonymous> (/tmp/.mount_Kit-Libizfr6/resources/app.asar/node_modules/x11/lib/xcore.js:88:21)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at Pipe.onStreamRead (node:internal/stream_base_commons:190:23) {
  error: 5,
  seq: 15,
  badParam: 0,
  minorOpcode: 0,
  majorOpcode: 17
}
```

**Reply by johnlindquist** [reply]
@TAKANOME-DEV I just tried googling some errors. Does this help?

https://github.com/flathub/org.gimp.GIMP/issues/23#issuecomment-394903354

**Reply by takanome-dev** [reply]
I didn't quite understand the solution describe in there, I'm not sure if my issue is related to that. But it finally works 🥳
Here is what I did:
- Reboot the machine
- Delete `~./kit`
- Follow the instructions mentioned above https://github.com/johnlindquist/kit/discussions/121#discussioncomment-4752431
- Download [Kit-Linux-1.40.62-x86_64.AppImage](https://github.com/johnlindquist/kitapp/releases/download/v1.40.62/Kit-Linux-1.40.62-x86_64.AppImage) and [run it](https://itsfoss.com/cant-run-appimage-ubuntu) 🎉

> **Note**
>
> Sometimes when using the shortcut to open the prompt, it won't show up and you will get the following message in the logs:
> ```sh
> Security: failed to start Application launched by gsd-media-keys.
> Sender: systemd
> Time: [time]
> Message: Failed to start Application launched by gsd-media-keys.
> Audit Session: 4 (not sure if it will be always the same number)
> Priority: 3  (not sure if it will be always the same number)
> Subject: A start job for unit UNIT has failed
> Defined By: systemd
>
> A start job for unit UNIT has finished with a failure.
> The job identifier is [some number] and the job result is failed.
> ```
> Not sure if it's only on me, but killing the process and restarting it will fix it :v: 

Thanks @johnlindquist

**Comment by geo242** [comment]
When I use the AppImage, am I supposed to manually install node into `~/.knode` beforehand? Looking at the logs, it appears that it's expecting node to be installed there. 

I am getting this error .... 

```Unhandled Promise Rejection
Error: Install not verified...
    at /tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:501114
    at l (/tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:478683)
    at Generator.<anonymous> (/tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:480021)
    at Generator.next (/tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:479046)
    at vd (/tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:484675)
    at a (/tmp/.mount_Kit-LixhokU5/resources/app.asar/main.prod.js:2:484879)```

**Reply by johnlindquist** [reply]
@geo242 Yes, the app _should_ be doing that for you. But the manual steps above should take care of it if not.

**Reply by geo242** [reply]
Got it, thanks! I unzipped the node tar.gz file into ~/.knode and now it seems to be working correctly.

**Reply by johnlindquist** [reply]
@geo242 Can you check your logs for `Download node.js from xxx` and see where it tried to download node from during setup?

~/.config/Kit.app/logs/main.log
~/.kit/logs/kit.log

**Reply by geo242** [reply]
For sure. I found the error... 

```14:20:22.510 › install-node.sh   Configuration
> Version:  v16.17.1 (resolved from )
> Prefix:   node
> Platform: darwin
> Arch:     x64

> Tarball URL: https://nodejs.org/dist/v16.17.1/node-v16.17.1-darwin-x64.tar.gz
> Installing Node.js, please wait…

{
  stderr: 'tar: node: Cannot open: No such file or directory\n' +
    'tar: Error is not recoverable: exiting now\n' +
    'x Command failed (exit code 23): curl --silent --fail https://nodejs.org/dist/v16.17.1/node-v16.17.1-darwin-x64.tar.gz\n'
}
14:20:22.516 › Stored version: 0.0.0 -> Current version: 1.40.62
```

**Comment by johnlindquist** [comment]
Update: Heya Linux Friends 👋

I would love help testing the `npx` tool I put together here: https://github.com/johnlindquist/kit/discussions/1052

Once I get enough people to verify that it works, I'll merge it into the main app setup process.

Cheers! - John

**Reply by mstine** [reply]
Not only does it work, but it also seems to work much better than the manual process. I got a bunch of fun examples pre-installed when I started up. These were not present when I installed the `.k*` directories manually. Love it!

**Reply by geo242** [reply]
I got the following error when I ran using node 14 and npm 6 (I use nvm and didn't realize that was my current environment)

```file:///home/geoff/.npm/_npx/74446/lib/node_modules/@johnlindquist/install-kit/index.js:9
process.env.NODE_VERSION ||= "16.17.1"
                         ^^^

SyntaxError: Unexpected token '||='
    at Loader.moduleStrategy (internal/modules/esm/translators.js:145:18)
    at async link (internal/modules/esm/module_job.js:67:21)
```

Once I switched to node 16 and npm 8, it succeeded. Might be a good idea to check the version at startup if possible.

**Reply by takanome-dev** [reply]
It works perfectly fine :clap: Pretty straightforward installation and really love all the new options added  🎉
